Add ICU change docs (#36734)
dotnet/runtime: docs/design/features/globalization-icu.md

# .NET Globalization ICU on all runtimes

Author: [Santiago Fernandez Madero](https://github.com/safern)

> In .NET Core our globalization APIs behavior used different underlying libraries, in Unix, we used [International Components for Unicode (ICU)](http://site.icu-project.org/home) and in Windows, we used [National Language Support (NLS)](https://docs.microsoft.com/en-us/windows/win32/intl/national-language-support). This caused some behavior differences in a handful number of Globalization APIs:

Reviewer: ```suggestion
In the past, the .NET globalization APIs have used different underlying libraries on different platforms.  On Unix, it used [International Components for Unicode (ICU)](http://site.icu-project.org/home), and on Windows, it used [National Language Support (NLS)](https://docs.microsoft.com/en-us/windows/win32/intl/national-language-support). This resulted in some behavioral differences in a handful of globalization APIs when running applications on different platforms, in particular affecting:
```

- Cultures and culture data
- String casing
- String sorting and searching
- Sort keys
- String Normalization
- Internationalized Domain Names (IDN) support
- Time Zone display name on Linux

> 

Reviewer: ```suggestion
Starting with .NET 5, developers have more control over the underlying library utilized, enabling applications to avoid such differences across platforms.
```

## ICU on Windows

> Windows 10 May 2019 Update and later versions started including [icu.dll](https://docs.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-) as part of the OS, so in .NET 5.0 we changed the default behavior for Globalization APIs when running in Windows, which tries to load `icu.dll` and if we can load it, we use ICU as the underlying native APIs, if we fail to load it, we use the legacy behavior (NLS), this will be the case for older versions of Windows.

Reviewer: ```suggestion
Windows 10 May 2019 Update and later versions include [icu.dll](https://docs.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-) as part of the OS, and .NET 5 now uses that by default. When running on Windows, .NET 5 tries to load `icu.dll` and uses it for the globalization implementation if it's available.  If that library can't be found or loaded, such as when running on older versions of Windows, .NET 5 will fall back to NLS-based implementation.
```

Reviewer 2: > Windows 10 May 2019 Update and later versions include [icu.dll](https://docs.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-) as part of the OS, and .NET 5 

Windows started including the ICU from RS2 but we'll have the minimum requirements of the version shipped in 19H1.

> *Note: CurrentCulture, CurrentUICulture and CurrentRegion still use Windows OS APIs to honor user settings*

Reviewer: ```suggestion
*Note: Even when using ICU, the `CurrentCulture`, `CurrentUICulture`, and `CurrentRegion` members still use Windows OS APIs to honor user settings.*
```

> *Note: CurrentCulture, CurrentUICulture and CurrentRegion still use Windows OS APIs to honor user settings*

Reviewer 3: > *Note: CurrentCulture, CurrentUICulture and CurrentRegion still use Windows OS APIs to honor user settings*

Can you clarify this? Do you mean when using those in some API, we'll use NLS instead (that seems quite limiting, but if that's the design fine)? Or do you mean we will ask Windows for the culture/region to use, but then use ICU tables to do the job.

Reviewer 2: Those are user profile settings on Windows. I would rephrase this as `CurrentCulture, CurrentUICulture and CurrentRegion properties are still initialized the same way we used to initialize using the Windows Win32 APIs and not ICU APIs`

Reviewer 3: But the point being that if I use ICU (possibly my own ICU) then yes I do get ICU behavior for `string.Contains(... , CurrentCulture)` etc?

Author: Yeah, it is just the initialization routines to get the current culture info from the OS that uses NLS, but for all other operations you would use ICU.

Reviewer 2: that is right. the functionality is fully provided by using ICU. initializing current culture is mainly reading the culture name from the OS settings and will use ICU to initialize the culture and performing any operations with it.

### Using NLS instead of ICU

> Given that this is a change in the Globalization behavior, we added support for disabling this feature, so that applications can use legacy behavior which relies on: [National Language Support (NLS)](https://docs.microsoft.com/en-us/windows/win32/intl/national-language-support).
> 
> Applications can enable NLS mode by either of the following:

Reviewer: ```suggestion
Using ICU instead of NLS may result in behavioral differences with some globalization-related operations. To revert back to using NLS, a developer can opt-out of the ICU implementation. Applications can enable NLS mode by either of the following:
```

> 1. in project file:

Reviewer: ```suggestion
1. In the project file:
```

```xml
<ItemGroup>
  <RuntimeHostConfigurationOption Include="System.Globalization.UseNls" Value="true" />
</ItemGroup>
```

> 2. in `runtimeconfig.json` file:

Reviewer: ```suggestion
2. In the `runtimeconfig.json` file:
```

```json
{
  "runtimeOptions": {
     "configProperties": {
       "System.Globalization.UseNls": true
      }
  }
}
```

> 3. setting environment variable value `DOTNET_SYSTEM_GLOBALIZATION_USENLS` to `true` or `1`.

Reviewer: ```suggestion
3. Setting the environment variable `DOTNET_SYSTEM_GLOBALIZATION_USENLS` to the value `true` or `1` (without quotes).
```

> *Note: value set in project file or `runtimeconfig.json` has higher priority than the environment variable.*

Reviewer: ```suggestion
*Note: A value set in the project or in the `runtimeconfig.json` file takes precedence over the environment variable.*
```

## App-local ICU

> In order to provide Common Locale Data Repository (CLDR) and behavior customization, we introduced another feature for applications to be able to carry their own copy of ICU as part of their closure. 

Reviewer: ```suggestion
Each release of ICU may bring with it bug fixes as well as updated Common Locale Data Repository (CLDR) data describing the world's languages.  Moving between versions of ICU can thus subtly impact application behavior when it comes to globalization-related operations.  To help application developer's ensure consistency across all deployments, .NET starting with .NET 5 also enables applications on both Windows and Unix to carry and use their own copy of ICU.
```

> Applications can enable NLS mode by either of the following:

Reviewer: ```suggestion
Applications can opt-in to an app-local ICU implementation mode by either of the following:
```

> 1. in project file:

Reviewer: ```suggestion
1. In the project file:
```

```xml
<ItemGroup>
  <RuntimeHostConfigurationOption Include="System.Globalization.AppLocalIcu" Value="<suffix>:<version> or <version>" />
</ItemGroup>
```

> 2. in `runtimeconfig.json` file:

Reviewer: ```suggestion
2. In the `runtimeconfig.json` file:
```

```json
{
  "runtimeOptions": {
     "configProperties": {
       "System.Globalization.AppLocalIcu": "<suffix>:<version> or <version>"
      }
  }
}
```

> 3. setting environment variable value `DOTNET_SYSTEM_GLOBALIZATION_APPLOCALICU` to `<suffix>:<version>` or `<version>`. 

Reviewer: ```suggestion
3. Setting the environment variable `DOTNET_SYSTEM_GLOBALIZATION_APPLOCALICU` to the value `<suffix>:<version>` or `<version>` (without quotes). 
```

> `<suffix>`: this is optional in the config property, but this follows the public ICU packaging conventions as when building a custom ICU you can customize it to produce the lib names and exported symbol names to contain a suffix. i.e: `libicuucmyapp` where `myapp` is the suffix. This can't be greater than 35 chars in length for the config switch.

Reviewer: ```suggestion
`<suffix>`: Optional suffix of less than 36 characters in length, following the public ICU packaging conventions. When building a custom ICU, you can customize it to produce the lib names and exported symbol names to contain a suffix, e.g. `libicuucmyapp` where `myapp` is the suffix.
```

> `<version>`: this has to be a valid ICU version, i.e: 67.1. This version will be used to load the binaries and to get the exported symbols.

Reviewer: ```suggestion
`<version>`: A valid ICU version, e.g. 67.1. This version will be used to load the binaries and to get the exported symbols.
```

> To load ICU when the app-local switch is set, we use `NativeLibrary.TryLoad` api which does probing in different paths, first it tries to find the library in `NATIVE_DLL_SEARCH_DIRECTORIES` property which is created by the dotnet host based on the `deps.json` file for the app. More details [here](https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/default-probing)

Reviewer: ```suggestion
To load ICU when the app-local switch is set, .NET uses the [NativeLibrary.TryLoad](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.nativelibrary.tryload) method which probes multiple paths, first trying to find the library in the `NATIVE_DLL_SEARCH_DIRECTORIES` property which is created by the dotnet host based on the `deps.json` file for the app. More details [here](https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/default-probing)
```

> For self contained apps, the user doesn't really need to do anything special, other than making sure ICU is side by side in the APP directory, this is because for self-contained apps, it's work directory is by default in `NATIVE_DLL_SEARCH_DIRECTORIES`. 

Reviewer: ```suggestion
For self-contained apps, no special action is required by the user, other than making sure ICU is in the APP directory (for self-contained apps, the working directory defaults to `NATIVE_DLL_SEARCH_DIRECTORIES`).
```

> If you're consuming ICU via a NuGet package, this will work in framework-dependent apps as NuGet will resolve the native assets and include them in the `deps.json` file and in the output directory for the app under the `runtimes` dir, then we will load it from there.

Reviewer: ```suggestion
If you're consuming ICU via a NuGet package, this will work in framework-dependent apps as NuGet will resolve the native assets and include them in the `deps.json` file and in the output directory for the app under the `runtimes` dir, and .NET will load it from there.
```

Reviewer 3: Does `APP directory` (lower case?) mean "next to the app.dll" or "in the app.dll folder or below"?

> The tricky part comes whenever it is a framework-dependent app (not self contained) and ICU is consumed from a local build. The SDK doesn't yet have a feature for "loose" native binaries to land into `deps.json`: https://github.com/dotnet/sdk/issues/11373. 
> 
> However, there is a workaround by adding something like this to the csproj:

Reviewer: ```suggestion
For framework-dependent apps (not self contained) where ICU is consumed from a local build, additional steps must be taken. The .NET SDK doesn't yet have a feature for "loose" native binaries to be incorporated into `deps.json` (https://github.com/dotnet/sdk/issues/11373). Instead, a developer can enable this by adding additional information into the application's project file, e.g.:
```

```xml
<ItemGroup>
  <IcuAssemblies Include="icu\*.so*" />
  <RuntimeTargetsCopyLocalItems Include="@(IcuAssemblies)" AssetType="native" CopyLocal="true" DestinationSubDirectory="runtimes/linux-x64/native/" DestinationSubPath="%(FileName)%(Extension)" RuntimeIdentifier="linux-x64" NuGetPackageId="System.Private.Runtime.UnicodeData" />
</ItemGroup>
```

> Note that this will have to be done for all the ICU binaries for the supported runtimes. Also, the `NuGetPackageId` metadata in the `RuntimeTargetsCopyLocalItems` item group, needs to match a NuGet package that the project actually references, it can't just be a dummy NuGet package.

Reviewer: ```suggestion
Note that this will have to be done for all the ICU binaries for the supported runtimes. Also, the `NuGetPackageId` metadata in the `RuntimeTargetsCopyLocalItems` item group needs to match a NuGet package that the project actually references.
```

### macOS behavior

> `MacOS` has a different behavior for resolving dependent dynamic libraries from the load commands specified in the `match-o` file than the Linux loader. In the Linux loader, we could just load `libicudata` first, then `libicuuc` and last `libicui18n` in that order to satisfy ICU dependency graph.

Reviewer 3: capitalization is `macOS` throughout (just a nit but would have to be right in docs)

> `MacOS` has a different behavior for resolving dependent dynamic libraries from the load commands specified in the `match-o` file than the Linux loader. In the Linux loader, we could just load `libicudata` first, then `libicuuc` and last `libicui18n` in that order to satisfy ICU dependency graph.
> 
> However, in MacOS this doesn't work. When building ICU in MacOS, you by default get a dynamic library with these load commands in libicuuc for example:

Reviewer: ```suggestion
`macOS` has a different behavior for resolving dependent dynamic libraries from the load commands specified in the `match-o` file than the Linux loader. In the Linux loader, .NET can try `libicudata` first, then `libicuuc`, and last `libicui18n` in that order to satisfy ICU dependency graph. However, on macOS this doesn't work. When building ICU on macOS, you by default get a dynamic library with these load commands in libicuuc, e.g.:
```

```sh
~/ % otool -L /Users/santifdezm/repos/icu-build/icu/install/lib/libicuuc.67.1.dylib
/Users/santifdezm/repos/icu-build/icu/install/lib/libicuuc.67.1.dylib:
 libicuuc.67.dylib (compatibility version 67.0.0, current version 67.1.0)
 libicudata.67.dylib (compatibility version 67.0.0, current version 67.1.0)
 /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
 /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 902.1.0)
```

> These commands are just referencing the name of the dependent libraries for the other components of ICU, so the loader will do the search following the `dlopen` conventions. Which involve having these libraries in the system directories or setting the `LD_LIBRARY_PATH` env vars, or having ICU at the app level directory. So if you can't set `LD_LIBRARY_PATH` or make sure that ICU binaries are at the app level directory, you will need to do some extra work.

Reviewer: ```suggestion
These commands are just referencing the name of the dependent libraries for the other components of ICU, so the loader will do the search following the `dlopen` conventions, which involve having these libraries in the system directories or setting the `LD_LIBRARY_PATH` env vars, or having ICU at the app level directory. So if you can't set `LD_LIBRARY_PATH` or make sure that ICU binaries are at the app level directory, you will need to do some extra work.
```

There are some directives for the loader, like `@loader_path` which tells the loader to search for that dependency in the same directory as the binary with that load command. So there's 2 ways to achieve this:

> ##### instal_name_tool -change

Reviewer 3: ```suggestion
##### install_name_tool -change
```
